src/mainboard/pcengines/apu2/mainboard.c

```diff
@@ -194,6 +196,88 @@ static void config_gpio_mux(void)
 	}
 }
 
+static void set_dimm_info(uint8_t *spd, struct dimm_info *dimm)
+{
+	const int spd_capmb[8] = {  1,  2,  4,  8, 16, 32, 64,  0 };
+	const int spd_ranks[8] = {  1,  2,  3,  4, -1, -1, -1, -1 };
+	const int spd_devw[8]  = {  4,  8, 16, 32, -1, -1, -1, -1 };
+	const int spd_busw[8]  = {  8, 16, 32, 64, -1, -1, -1, -1 };
+
+	int capmb = spd_capmb[spd[SPD_DENSITY_BANKS] & 7] * 256;
+	int ranks = spd_ranks[(spd[SPD_ORGANIZATION] >> 3) & 7];
+	int devw  = spd_devw[spd[SPD_ORGANIZATION] & 7];
+	int busw  = spd_busw[spd[SPD_BUS_DEV_WIDTH] & 7];
+
+	switch (spd[12]) {
+		case 0x0a:
+			dimm->ddr_frequency = 1600;
+			break;
+		case 0x0c:
+			dimm->ddr_frequency = 1333;
+			break;
+		default:
+			dimm->ddr_frequency = 0;
+			break;
+	}
+	
+	dimm->ddr_type = MEMORY_TYPE_DDR3;
+
+	/* Parse the SPD data to determine the DIMM information */
+
+	dimm->dimm_size = capmb / 8 * busw / devw * ranks;  /* MiB */
+	dimm->mod_type = spd[3] & 0xf;
+	strncpy((char *)&dimm->module_part_number[0], (char *)&spd[0x80],
```

Comment:
@finus16 our SPD is 128 bytes and has no part number. What You are trying to do here is to copy a string outside of SPD. You should assign null character to dimm->module_part_number[0]

Reply:
fixed